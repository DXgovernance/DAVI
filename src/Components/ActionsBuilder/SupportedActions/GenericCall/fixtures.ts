import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'Components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const genericDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.GENERIC_CALL,
  function: ERC20Contract.getFunction('increaseAllowance'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    param1: BigNumber.from(10000),
    param2: BigNumber.from(10000),
    param3: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
