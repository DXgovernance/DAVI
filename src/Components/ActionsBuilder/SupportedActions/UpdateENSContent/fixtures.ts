import { BigNumber, utils } from 'ethers';
import { DecodedCall } from 'Components/ActionsBuilder/types';
import { SupportedAction } from 'Components/ActionsBuilder/types';
import ENSPublicResolver from 'contracts/ENSPublicResolver.json';

const ENSPublicResolverContract = new utils.Interface(ENSPublicResolver.abi);

export const mockDecodedCallUpdateENSContent: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ENS_UPDATE_CONTENT,
  function: ENSPublicResolverContract.getFunction('setContenthash'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    node: '0x0000000000000000000000000000000000000000',
    hash: '0x0000000000000000000000000000000000000000',
  },
  optionalProps: {
    ensName: 'test',
    ipfsHash: 'Qm',
  },
};
