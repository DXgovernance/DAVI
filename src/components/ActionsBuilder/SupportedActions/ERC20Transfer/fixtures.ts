import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const erc20TransferEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const erc20TransferDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    _to: '0xaF8eB8C3A5d9d900AA0B98e3Df0bcC17d3C5F698',
    _value: BigNumber.from('5000000000000000000'),
  },
};
