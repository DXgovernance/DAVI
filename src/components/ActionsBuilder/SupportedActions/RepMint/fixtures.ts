import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import ERC20SnapshotRep from 'contracts/ERC20SnapshotRep.json';

const ERC20SnapshotRepContract = new utils.Interface(ERC20SnapshotRep.abi);

export const repMintEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '',
  value: BigNumber.from(0),
  args: {
    to: '',
    amount: BigNumber.from(0),
  },
};

export const repMintDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
    amount: BigNumber.from(10000),
  },
};
