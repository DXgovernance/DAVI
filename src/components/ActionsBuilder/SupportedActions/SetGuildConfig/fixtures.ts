import { BigNumber, utils } from 'ethers';
import { SupportedAction } from 'components/ActionsBuilder/types';
import BaseERC20Guild from 'contracts/BaseERC20Guild.json';

export const mockBigNumber = BigNumber.from(100000000);
export const BaseERC20GuildContract = new utils.Interface(BaseERC20Guild.abi);

export const emptyDecodedCall = {
  from: '',
  callType: SupportedAction.SET_GUILD_CONFIG,
  function: BaseERC20GuildContract.getFunction('setConfig'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _proposalTime: '',
    _timeForExecution: '',
    _votingPowerPercentageForProposalExecution: '',
    _votingPowerPercentageForProposalCreation: '',
    _voteGas: '',
    _maxGasPrice: '',
    _maxActiveProposals: '',
    _lockTime: '',
    _minimumMembersForProposalCreation: '',
    _minimumTokensLockedForProposalCreation: '',
  },
  optionalProps: {},
};
