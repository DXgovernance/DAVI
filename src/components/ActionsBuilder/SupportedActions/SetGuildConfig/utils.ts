import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
