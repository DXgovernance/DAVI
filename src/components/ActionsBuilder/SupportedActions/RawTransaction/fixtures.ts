import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';

export const rawDataCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.RAW_TRANSACTION,
  function: null,
  to: MOCK_ADDRESS,
  value: BigNumber.from('1000000000000000000'),
  args: null,
  optionalProps: {
    data: '0x01',
  },
};
