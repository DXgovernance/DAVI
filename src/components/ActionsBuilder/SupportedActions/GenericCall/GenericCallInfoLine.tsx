import { BigNumber } from 'ethers';
import useRichContractData from 'hooks/Guilds/contracts/useRichContractData';
import { RichContractFunctionParam } from 'hooks/Guilds/contracts/useRichContractRegistry';
import useBigNumberToString from 'hooks/Guilds/conversions/useBigNumberToString';
import { Interweave } from 'interweave';
import {
  useERC20Info,
  getTokenInfoParsedParams,
} from 'hooks/Guilds/erc20/useERC20Info';
import { Segment } from '../common/infoLine';
import { useMemo } from 'react';
import { FiArrowRight, FiCode } from 'react-icons/fi';
import { ActionViewProps } from '..';
import GenericCallParamsMatcher from './GenericCallParamsMatcher';
import { useTranslation } from 'react-i18next';
import { preventEmptyString } from 'utils';

export interface FunctionParamWithValue extends RichContractFunctionParam {
  value: string;
}

const GenericCallInfoLine: React.FC<ActionViewProps> = ({
  decodedCall,
  approveSpendTokens,
  compact = false,
}) => {
  const { t } = useTranslation();
  const { data: tokenInfo } = useERC20Info(approveSpendTokens?.token);
  const approvalAmount = useBigNumberToString(
    preventEmptyString(approveSpendTokens?.amount),
    tokenInfo?.decimals
  );
  const { functionData } = useRichContractData(decodedCall);

  function getReadableStringForParam(type: string, value: any, name: string) {
    if (!type || !value) return null;

    if (name === 'proposalId') {
      return `${value.substring(0, 6)}...${value.substring(value.length - 4)}`;
    }

    if (type.startsWith('uint') || type.startsWith('int')) {
      return BigNumber.from(value).toString();
    }

    return value;
  }

  const params: FunctionParamWithValue[] = useMemo(() => {
    if (!decodedCall || !functionData) return null;
    return functionData.params
      .map(param => ({
        ...param,
        value: getReadableStringForParam(
          param.type,
          decodedCall.args[param.name],
          param.name
        ),
      }))
      .concat(getTokenInfoParsedParams(tokenInfo));
  }, [functionData, decodedCall, tokenInfo]);

  return (
    <>
      <Segment>
        <FiCode size={16} />
      </Segment>

      {compact || !params ? (
        <>
          {!!approveSpendTokens && (
            <>
              <Segment>
                {!!tokenInfo
                  ? approvalAmount + tokenInfo?.symbol
                  : t('unknownToken')}
              </Segment>
              <Segment>
                <FiArrowRight />
              </Segment>
            </>
          )}
          <Segment>
            {decodedCall?.richFunctionData?.title ||
              decodedCall?.function?.name}
          </Segment>
        </>
      ) : (
        <>
          {!!approveSpendTokens && (
            <>
              <Segment>
                {!!tokenInfo
                  ? approvalAmount + tokenInfo?.symbol ?? ''
                  : 'Unknown token'}
              </Segment>
              <Segment>
                <FiArrowRight />
              </Segment>
            </>
          )}
          <Segment>
            <Interweave
              content={functionData?.templateLiteral}
              matchers={[
                new GenericCallParamsMatcher('genericCallParamsMatcher', {
                  params,
                }),
              ]}
            />
          </Segment>
        </>
      )}
    </>
  );
};

export default GenericCallInfoLine;
