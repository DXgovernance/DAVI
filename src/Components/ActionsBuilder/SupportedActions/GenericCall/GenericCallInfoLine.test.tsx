import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import { approvetokenSpendingMock, genericDecodedCallMock } from './fixtures';
import GenericCallInfoLine from './GenericCallInfoLine';
import { RichContractFunction } from 'hooks/Guilds/contracts/useRichContractRegistry';

jest.mock('hooks/Guilds/contracts/useRichContractData', () => ({
  __esModule: true,
  default: (): { functionData: RichContractFunction } => ({
    functionData: {
      title: 'Test Function',
      functionName: 'TestFunction',
      shortDescription: 'Test function description',
      longDescription: 'This is a long description for this test function.',
      params: [
        {
          name: 'param1',
          type: 'uint256',
          component: 'TestComponent',
          description: 'Test parameter 1',
          defaultValue: '',
        },
        {
          name: 'param2',
          type: 'int256',
          component: 'TestComponent',
          description: 'Test parameter 2',
          defaultValue: '',
        },
        {
          name: 'param3',
          type: 'address',
          component: 'TestComponent',
          description: 'Test parameter 3',
          defaultValue: '',
        },
      ],
      // eslint-disable-next-line no-template-curly-in-string
      templateLiteral: 'Test template ${param1} ${param2} ${param3}',
      spendsTokens: true,
    },
  }),
}));

jest.mock('hooks/Guilds/ether-swr/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

const mockBigNumber = BigNumber.from(100000000);
jest.mock('hooks/Guilds/ether-swr/erc20/useERC20Info', () => ({
  useERC20Info: () => ({
    name: 'Test ERC20',
    symbol: 'TEST',
    decimals: 18,
    totalSupply: mockBigNumber,
  }),
  getTokenInfoParsedParams: () => [],
}));

describe('GenericCallInfoLine', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <GenericCallInfoLine decodedCall={genericDecodedCallMock} />
    );
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot in compact mode', () => {
    const { container } = render(
      <GenericCallInfoLine decodedCall={genericDecodedCallMock} compact />
    );
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot with token approval', () => {
    const { container } = render(
      <GenericCallInfoLine
        decodedCall={genericDecodedCallMock}
        approveSpendTokens={approvetokenSpendingMock}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
