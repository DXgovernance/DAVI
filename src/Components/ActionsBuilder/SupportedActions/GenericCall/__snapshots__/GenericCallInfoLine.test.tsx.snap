// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericCallInfoLine Should match snapshot 1`] = `
.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
   
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="16 18 22 12 16 6"
      />
      <polyline
        points="8 6 2 12 8 18"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <span>
      Test template 
      <span
        matchedparam="param1"
      >
        10000
      </span>
       
      <span
        matchedparam="param2"
      >
        10000
      </span>
       
      <span
        matchedparam="param3"
      >
        0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351
      </span>
    </span>
  </span>
   
</div>
`;

exports[`GenericCallInfoLine Should match snapshot in compact mode 1`] = `
.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
   
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="16 18 22 12 16 6"
      />
      <polyline
        points="8 6 2 12 8 18"
      />
    </svg>
  </span>
   
</div>
`;

exports[`GenericCallInfoLine Should match snapshot with token approval 1`] = `
.c0 {
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
   
  <span
    class="c0"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="16 18 22 12 16 6"
      />
      <polyline
        points="8 6 2 12 8 18"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <span>
      Test template 
      <span
        matchedparam="param1"
      >
        10000
      </span>
       
      <span
        matchedparam="param2"
      >
        10000
      </span>
       
      <span
        matchedparam="param3"
      >
        0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351
      </span>
    </span>
  </span>
   
</div>
`;
